x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/action/TransportFreezeIndexAction.java

```diff
@@ -0,0 +1,198 @@
+            new AckedClusterStateUpdateTask<AcknowledgedResponse>(Priority.URGENT, request, listener) {
+            @Override
+            public ClusterState execute(final ClusterState currentState) {
+                final MetaData.Builder builder = MetaData.builder(currentState.metaData());
+                ClusterBlocks.Builder blocks = ClusterBlocks.builder().blocks(currentState.blocks());
+                for (Index index : concreteIndices) {
+                    IndexMetaData meta = currentState.metaData().index(index);
+                    if (meta.getState() != IndexMetaData.State.CLOSE) {
+                        throw new IllegalStateException("index [" + index.getName() + "] is not closed");
+                    }
+                    final IndexMetaData.Builder imdBuilder = IndexMetaData.builder(meta);
+                    final Settings.Builder settingsBuilder =
+                        Settings.builder()
+                            .put(currentState.metaData().index(index).getSettings())
+                            .put("index.blocks.write", request.freeze())
+                            .put(FrozenEngine.INDEX_FROZEN.getKey(), request.freeze())
+                            .put(IndexSettings.INDEX_SEARCH_THROTTLED.getKey(), request.freeze());
+                    if (request.freeze()) {
+                        blocks.addIndexBlock(index.getName(), IndexMetaData.INDEX_WRITE_BLOCK);
+                        // we never remove this block when unfreeze for now. we don't know if it was read-only
+                    }
+                    imdBuilder.settings(settingsBuilder);
+                    builder.put(imdBuilder.build(), true);
+                }
+                return ClusterState.builder(currentState).blocks(blocks).metaData(builder).build();
+            }
+
+            @Override
+            protected AcknowledgedResponse newResponse(boolean acknowledged) {
+                return new AcknowledgedResponse(acknowledged);
+            }
+        });
+    }
+
+    @Override
+    protected ClusterBlockException checkBlock(FreezeRequest request, ClusterState state) {
+        return state.blocks().indicesBlockedException(ClusterBlockLevel.METADATA_WRITE,
+            indexNameExpressionResolver.concreteIndexNames(state, request));
+    }
+
+    public static class FreezeIndexAction extends Action<AcknowledgedResponse> {
+
+        public static final FreezeIndexAction INSTANCE = new FreezeIndexAction();
+        public static final String NAME = "indices:admin/freeze";
+
+        private FreezeIndexAction() {
+            super(NAME);
+        }
+
+        @Override
+        public AcknowledgedResponse newResponse() {
+            return new AcknowledgedResponse();
+        }
+    }
+
+    public static class FreezeRequest extends AcknowledgedRequest<FreezeRequest>
+        implements IndicesRequest.Replaceable {
+        private OpenIndexRequest openIndexRequest;
```

Comment:
Why should we embed an `OpenIndexRequest` into a `FreezeRequest`? I think it only adds a strong coupling between the two without much benefit. We have many request that contain indices and indices options, maybe we could put this into an AcknowledgedIndicesRequest and have Free/Open/Delete (and maybe later Close) request extend it?

Reply:
this is a leftover. good catch